### DISCUSSION TITLE
How to build SMARTS programmatically?

### DISCUSSION THREAD

**Discussion by simonmb** [rdkit/rdkit#4683] [Q&A] [answered]
I am looking for a way to build SMARTS programmatically. To add details to the query by code.
How could this be done?

It is possible to add a recursive query like so:

```python
patt = Chem.MolFromSmarts('[CH2]')
print(rdmolfiles.MolFragmentToSmarts(patt, [0])) # returns '[C&H2]'

rdmolops.AddRecursiveQuery(patt, Chem.MolFromSmarts('C=C'), 0)

print(rdmolops.AddRecursiveQuery(patt, Chem.MolFromSmarts('C=C'), 0)) # returns '[C&H2&$(C=C)]'
```

But how would you add the information about the degree of an atom to the pattern? Or the amount of implicit/explicit hydrogens? Or if the atom is aromatic?

**Comment by greglandrum** [comment, marked as answer]
Hi @simonmb : the module `rdqueries` has the tools that you need for this. Here is a simple example:
```
In [2]: from rdkit import Chem

In [3]: m = Chem.MolFromSmarts('CC')

In [4]: from rdkit.Chem import rdqueries

In [5]: at = m.GetAtomWithIdx(0)

In [6]: q = rdqueries.ExplicitDegreeEqualsQueryAtom(3)

In [7]: m2 = Chem.RWMol(m)

In [8]: m2.ReplaceAtom(0,q)

In [9]: Chem.MolToSmarts(m2)
Out[9]: '[D3]C'

In [10]: q.ExpandQuery(rdqueries.AtomNumEqualsQueryAtom(6))

In [11]: m2.ReplaceAtom(0,q)

In [12]: Chem.MolToSmarts(m2)
Out[12]: '[D3&#6]C'

```
There are a bunch of different kinds of query types available in `rdqueries`.

You can also use atoms from molecules constructed from SMARTS:
```
In [13]: m3 = Chem.MolFromSmarts('[a]')

In [14]: q.ExpandQuery(m3.GetAtomWithIdx(0))

In [15]: m2.ReplaceAtom(0,q)

In [16]: Chem.MolToSmarts(m2)
Out[16]: '[D3&#6&a]C'
```

or atoms constructed from SMARTS:
```
In [18]: a = Chem.AtomFromSmarts('[H1]')

In [19]: q.ExpandQuery(a)

In [20]: m2.ReplaceAtom(0,q)

In [21]: Chem.MolToSmarts(m2)
Out[21]: '[D3&#6&a&H1]C'
```

But I suspect that `rdqueries` will be more useful.

**Reply by simonmb** [reply]
Thank you, I don't know how I didn't find this myself.

**Reply by greglandrum** [reply]
oh, that part of the code is definitely "underdocumented"... it's easy to miss.

**Reply by simonmb** [reply]
Is it possible to enumerate the queries on a specific QueryAtom?

**Reply by greglandrum** [reply]
No. You can call `atom.DescribeQuery()`, but that just provides you with text.

**Reply by simonmb** [reply]
Yes I know, once you told me to parse it because I needed information about the query. Would it be a lot of work to expose this to python? That would be the first part of SMARTS matching.

My idea was to generate all structures the SMARTS you want to compare would match and then match the structures. Basically whenever you have an OR you split the structures being create into two. Or maybe have a C++ function that would generate said structures and expose this to python. But my understandings of SMARTS are limited and maybe this is way more difficult than I think. :)

**Reply by simonmb** [reply]
Sorry, I came across the need to specify the type of bond any: ~
What is the rdkit.Chem.rdchem.BondType type I should use? I tried with OTHER and UNSPECIFIED but neither would create a SMARTS pattern with the bond type any ~.

**Reply by greglandrum** [reply]
There isn’t a simple bond type to correspond to the “null bond” query bond type.

**Reply by mwojcikowski** [reply]
I think the closest you can get is by negating the other bond orders `!-!=!:!#`

**Reply by simonmb** [reply]
This is the inconsistency I find:
```python
patt1 = Chem.MolFromSmarts('[C]~[C]')
print(patt1.GetBonds()[0].GetBondType()) # UNSPECIFIED
patt2 = Chem.MolFromSmarts('[C][C]')
print(patt2.GetBonds()[0].GetBondType()) # SINGLE

patt3 = Chem.MolFromSmarts('[C]')
patt3 = Chem.RWMol(patt3)
patt3.AddAtom(Chem.AtomFromSmarts('[C]'))
patt3.AddBond(0, 1, rdchem.BondType.UNSPECIFIED)

print(Chem.MolToSmarts(patt1)) # C~C 
print(Chem.MolToSmarts(patt2)) # CC
print(Chem.MolToSmarts(patt3)) # CC
```
I would have expected the SMARTS of patt1 and patt 3 to be identical. Or is this a bug?

**Reply by mwojcikowski** [reply]
`patt2='[C][C]'` is strictly speaking equal to `[C]-,:[C]`

**Reply by simonmb** [reply]
But what about the difference between patt1 and patt3?
